File: design-pattern/factory-pattern/factory-method/factory_method.cpp
#include <iostream>
using namespace std;

class Burger
{
public:
    virtual void prepareBurger() = 0;
    virtual ~Burger() {}
};

class BasicBurger : public Burger
{
public:
    void prepareBurger() override
    {
        cout << "Preparing the basic burger" << endl;
    }
};

class StandardBurger : public Burger
{
public:
    void prepareBurger() override
    {
        cout << "Preparing the standard burger" << endl;
    }
};

class PremiumBurger : public Burger
{
public:
    void prepareBurger() override
    {
        cout << "Preparing the premium burger" << endl;
    }
};

class BasicWheatBurger : public Burger
{
public:
    void prepareBurger() override
    {
        cout << "preparing the basic WheatBurger" << endl;
    }
};

class StandardWheatBurger : public Burger
{
public:
    void prepareBurger() override
    {
        cout << "preparing the standard WheatBurger" << endl;
    }
};

class PremiumWheatBurger : public Burger
{
public:
    void prepareBurger() override
    {
        cout << "preparing the premium WheatBurger" << endl;
    }
};

class BurgerFactory
{
public:
    virtual Burger *createBurger(string type) = 0;
    virtual ~BurgerFactory() {}
};

class SimpleBurgerFactory : public BurgerFactory
{
public:
    Burger *createBurger(string type) override
    {
        if (type == "basic")
            return new BasicBurger();
        else if (type == "standard")
            return new StandardBurger();
        else if (type == "premium")
            return new PremiumBurger();
        return nullptr;
    }
};

class WheatBurgerFactory : public BurgerFactory
{
public:
    Burger *createBurger(string type) override
    {
        if (type == "basic")
            return new BasicWheatBurger();
        else if (type == "standard")
            return new StandardWheatBurger();
        else if (type == "premium")
            return new PremiumWheatBurger();
        return nullptr;
    }
};

int main()
{
    BurgerFactory *factory = new WheatBurgerFactory();
    string type = "premium";

    Burger *myBurger = factory->createBurger(type);
    if (myBurger)
        myBurger->prepareBurger();

    delete myBurger;
    delete factory;

    return 0;
}

File: design-pattern/factory-pattern/factory-method/factory_method.ts
interface Burger {
    prepareBurger(): void
};

class BasicBurger implements Burger {
    prepareBurger(): void {
        console.log("cooking basic burger");
    };
}

class StandardBurger implements Burger {
    prepareBurger(): void {
        console.log("cooking the standard burger");
    };
};

class PremiumBurger implements Burger {
    prepareBurger(): void {
        console.log("cooking the Premium burger")
    }
};

class BasicWheatBurger implements Burger {
    prepareBurger(): void {
        console.log("preparing the basic wheat burger");
    };
};

class StandardWheatBurger implements Burger {
    prepareBurger(): void {
        console.log("preparing the standard wheat burger")
    };
};

class PrepareWheatBurger implements Burger {
    prepareBurger(): void {
        console.log("preparing the premium wheat burger")
    };
};

interface burgerFactory {
    createBurger(type: string): Burger | null
};

class planBurger implements burgerFactory {
    createBurger(type: string): Burger | null {
        if (type == "basic") {
            return new BasicBurger();
        } else if (type == "standard") {
            return new StandardBurger();
        } else if (type == "premium") {
            return new PremiumBurger();
        } else {
            return null;
        }
    }
}

class wheatBurger implements burgerFactory {
    createBurger(type: string): Burger | null {
        if (type == "basic") {
            return new BasicWheatBurger();
        } else if (type == "standard") {
            return new StandardWheatBurger();
        } else if (type == "premium") {
            return new PrepareWheatBurger();
        } else {
            return null;
        }
    }
};


const burgerFactory = new wheatBurger();
let type: string = "premium";
const burgerTry = burgerFactory.createBurger(type);
burgerTry?.prepareBurger();